src/util/block_tools.py

```diff
@@ -1118,3 +1102,114 @@ def unfinished_block_to_full_block(
+            full_block,
+            required_iters,
+        )
+        height_to_hash[uint32(full_block.height)] = full_block.header_hash
+    return difficulty
+
+
+def get_icc(
+    constants,
+    vdf_end_total_iters: uint128,
+    finished_sub_slots: List[EndOfSubSlotBundle],
+    latest_sub_block: SubBlockRecord,
+    sub_blocks: Dict[bytes32, SubBlockRecord],
+    sub_slot_start_total_iters: uint128,
+) -> Tuple[Optional[VDFInfo], Optional[VDFProof]]:
+    if latest_sub_block.deficit == constants.MIN_SUB_BLOCKS_PER_CHALLENGE_BLOCK:
+        # Curr block has deficit either 4 or 5 so no need for ICC vdfs
+        return None, None
+    else:
+        if len(finished_sub_slots) == 0:
+            curr = latest_sub_block  # curr deficit is 0, 1, 2, 3, or 4
+            while curr.deficit < constants.MIN_SUB_BLOCKS_PER_CHALLENGE_BLOCK - 1 and not curr.first_in_sub_slot:
+                curr = sub_blocks[curr.prev_hash]
+            icc_iters = uint64(vdf_end_total_iters - latest_sub_block.total_iters)
+            icc_input = latest_sub_block.infused_challenge_vdf_output
+            if curr.deficit == constants.MIN_SUB_BLOCKS_PER_CHALLENGE_BLOCK - 1:  # Deficit 4
+                icc_challenge_hash = curr.challenge_block_info_hash
+                icc_input = ClassgroupElement.get_default_element()
+            else:
+                # First sub block in sub slot has deficit 0,1,2 or 3
+                icc_challenge_hash = curr.finished_infused_challenge_slot_hashes[-1]
+        else:
+            icc_challenge_hash = finished_sub_slots[-1].infused_challenge_chain.get_hash()
+            icc_iters = uint64(vdf_end_total_iters - sub_slot_start_total_iters)
+            icc_input = ClassgroupElement.get_default_element()
+
+    return get_vdf_info_and_proof(
+        constants,
+        icc_input,
+        icc_challenge_hash,
+        icc_iters,
+    )
+
+
+def handle_end_of_epoch(
+    constants, new_slot, difficulty, deficit, ips, sp_iters, sub_blocks, prev_block, height_to_hash, curr_epoch
+):
+    new_epoch = False
+    if len(sub_blocks.keys()) == constants.EPOCH_SUB_BLOCKS * (curr_epoch + 1):
+        new_epoch = True
+        # new difficulty
+        difficulty = get_next_difficulty(
+            constants,
+            sub_blocks,
+            height_to_hash,
+            prev_block.header_hash,
+            prev_block.height,
+            deficit,
+            difficulty,
+            new_slot,
+            sp_iters,
+        )
+
+        # new iterations per slot
+        ips = get_next_ips(
+            constants,
+            height_to_hash,
+            sub_blocks,
+            prev_block.header_hash,
+            prev_block.height,
+            deficit,
+            ips,
+            new_slot,
+            sp_iters,
+        )
+
+    return ips, difficulty, new_epoch
+
+
+def handle_end_of_sub_epoch(
+    prev_block, sub_blocks, prev_subepoch_summary_hash, overflow_count, difficulty, ips, curr_sub_epoch
```

Comment:
Pass in constants instead of hardcoding to 384. Also use types for function arguments

Reply:
sure, next time just leave it and ill take care of it, i wanted to commit and push the logic before the days end